### ISSUE TITLE
CLI options could be grouped and otherwise presented better

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#3612]
The output of `eslint --help` is pretty hard to read with the number of options we now have, whether or not the options have short or long descriptions.  Plus, the options are not in a clear order that anyone on the project could easily perceive, understand, and explain.

It might be a good idea to have a discussion on how we can improve the options listing.

I believe the single biggest improvement we could make is to group related options under separate headers. This will result in natural whitespace between groups, which would allow us to slightly relax UX standards for the individual option descriptions themselves. In addition, we could revisit some of the descriptions if necessary.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
@platinumazure can you propose what you think would make sense?

**Comment by platinumazure** [comment]
Very roughly, here's how I might group the options (subject to further tweaking):

```text
eslint [options] file.js [file.js] [dir]

Basic configuration:
  -c, --config path::String   Use configuration from this file or sharable config
  --no-eslintrc               Disable use of configuration from .eslintrc
  --env [String]              Specify environments
  --ext [String]              Specify JavaScript file extensions - default: .js
  --global [String]           Define global variables
  --parser String             Specify the parser to be used - default: espree

Specifying rules and plugins:
  --rulesdir [path::String]   Use additional rules from this directory
  --plugin [String]           Specify plugins
  --rule Object               Specify rules

Using stdin:
  --stdin                     Lint code provided on <STDIN> - default: false
  --stdin-filename String     Specify filename to process STDIN as

Ignoring files:
  --ignore-path path::String  Specify path of ignore file
  --no-ignore                 Disable use of .eslintignore
  --ignore-pattern String     Pattern of files to ignore (in addition to those in .eslintignore)

Handling warnings:
  --quiet                     Report errors only - default: false
  --max-warnings Number       Number of warnings to trigger nonzero exit code - default: -1

Output:
  -f, --format String         Use a specific output format - default: stylish
  --no-color                  Disable color in piped output
  -o, --output-file path::String  Specify file to write report to

Miscellaneous:
  --init                      Run config initialization wizard - default: false
  -h, --help                  Show help
  -v, --version               Outputs the version number
```

Obviously some tweaking may be needed, but I think that conveys the concept pretty well.

**Comment by gyandeeps** [comment]
Can this be achieved using the current module for options?

**Comment by platinumazure** [comment]
I should think so. Take a look at `lib/options.js` and you'll see that optionator is configured with an array; most options have option objects, but the very first thing in the array is a header. I believe we can achieve the subsections by adding more header objects in the configuration and moving the options around to fit the appropriate groupings (whether that be what I suggested above, or something else).

**Comment by IanVS** [comment]
I'd like to see a `Config Bootstrap` or `Project Initialization` or similar heading which would contain `--init` and eventually `--autoconfig`

**Comment by IanVS** [comment]
Also, I really like this idea of grouping options under headers.

**Comment by platinumazure** [comment]
@IanVS Sure, why not? That's a good idea and could be done easily. I just couldn't think of a heading for it but what you suggested sounds great.

**Comment by nzakas** [comment]
@platinumazure looks good, do you want to investigate more.

@IanVS we can consider that after autoconfig is added, as I'm not sure it should be separate from `--init ` yet. Let's just stick with the current options for now.

**Comment by platinumazure** [comment]
@nzakas I can submit a PR for the above configuration to get things rolling, then we can discuss in there if we can fix grouping, ordering of groups, or ordering within groups. I see you've thrown an accepted label so I assume you have no objections-- let me know if I have misunderstood.

**Comment by nzakas** [comment]
No objections - your suggestion looks like a great starting point.

**Comment by platinumazure** [comment]
Thanks. One question about coding conventions-- could you please look at `lib/options.js` and tell me which style of array-of-objects you prefer?

Style 1 (first one-third of the options array):

```js
// ...
}, {
    key: "value"
}, {
    key: "value2"
}
```

Style 2 (further down in the same array):

```js
    // ...
    {
        key: "value"
    },
    {
        key: "value"
    }
```

Thanks!

**Comment by nzakas** [comment]
I'd prefer the latter.

**Comment by platinumazure** [comment]
Understood.

**Comment by platinumazure** [comment]
One more question for the group: Should we keep `docs/user-guide/command-line-interface.md` *exactly* in sync with the options?

Or, more specifically:
1. Should I group the options there, in the same way as I am doing for `lib/options.js`?
2. Do we really want exactly the optionator text there, when perhaps it might be sufficient to have just the options and the basic description there and let the rest of the page flesh out details and defaults?

**Comment by nzakas** [comment]
> Should we keep docs/user-guide/command-line-interface.md exactly in sync with the options?

Yes.

> Should I group the options there, in the same way as I am doing for lib/options.js?

Yes

> Do we really want exactly the optionator text there

Yes
